### PR TITLE
Handler for known LoweringErrors.

### LINKED ISSUE #2993: Invalid use of `make_function` needs catching before lowering
e.g. https://github.com/numba/numba/issues/2992 ends up as `make_function`+`LoweringError`, this particular path ought to be caught earlier and rewritten as a friendly user message as it is almost always down to use of an unsupported construct.

### PR DESCRIPTION
**Pull request by stuartarchibald**
Quite a few `LoweringError`s are from the use of known and detectable
unsupported language features. This adds a stage to the compiler
pipeline just ahead of the lowering pass that will try and identify
such cases and raise a more helpful error message.

Fixes #2993

### CODE REVIEW

**Comment by ehsantn** [comment]
Why is a new pass needed? Can't the Lowerer throw this error when it sees `make_function`?

File: numba/ir_utils.py [hunk 1]
```diff
@@ -1682,3 +1682,44 @@ def is_namedtuple_class(c):
...
+def raise_on_unsupported_feature(func_ir):
+    """
+    Helper function to walk IR and raise if it finds op codes
+    that are unsupported. Could be extended to cover IR sequences
+    as well as op codes. Intended use is to call it as a pipeline
+    stage just prior to lowering to prevent LoweringErrors for known
+    unsupported features.
+    """
+    for blk in func_ir.blocks.values():
+        for stmt in blk.body:
```
Comment (sklam) [inline numba/ir_utils.py]: this loop can be simplified with `blk.find_exprs()`

**Comment by stuartarchibald** [comment]
Following discussion in yesterday's Numba core dev meeting. The new pass remains so as to keep a separation of concerns between lowering, which should lower any valid IR, and validation. The pass is renamed `stage_ir_legalization` to reflect the use based on a similar pass in LLVM.

File: numba/ir_utils.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/ir_utils.py]: Yes, thanks, I always for get that! Fixed.

**Pull request merged**